### PR TITLE
Move all torch.nn.modules type annotations inline

### LINKED ISSUE #39463: JIT test suite is nondeterministic
Steps to reproduce:
1. Induce a bug by adding more type annotations. https://github.com/ezyang/pytorch/tree/poc/jit-bug is what I was working on when this happened
2. Run these tests:

```
(/home/ezyang/local/pytorch-tmp-env) [ezyang@devvm066.ash0 ~/local/pytorch-tmp] python test/test_jit.py TestScript.test_circular_depend
ency
Couldn't download test skip set, leaving all tests enabled...
.
----------------------------------------------------------------------
Ran 1 test in 0.211s

OK
(/home/ezyang/local/pytorch-tmp-env) [ezyang@devvm066.ash0 ~/local/pytorch-tmp] python test/test_jit.py TestScript.test_attribute_in_init TestScript.test_circular_dependency
Couldn't download test skip set, leaving all tests enabled...
.F
======================================================================
FAIL: test_circular_dependency (__main__.TestScript)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/test_jit.py", line 5303, in test_circular_dependency
    self.getExportImportCopy(C())
  File "/data/users/ezyang/pytorch-tmp/torch/jit/__init__.py", line 1581, in init_then_script
    original_init(self, *args, **kwargs)
  File "test/test_jit.py", line 5296, in __init__
    self.foo = torch.nn.Sequential(B())
  File "/data/users/ezyang/pytorch-tmp/torch/jit/__init__.py", line 1587, in init_then_script
    self.__dict__["_actual_script_module"] = torch.jit._recursive.create_script_module(self, make_stubs)
  File "/data/users/ezyang/pytorch-tmp/torch/jit/_recursive.py", line 305, in create_script_module
    concrete_type = concrete_type_store.get_or_create_concrete_type(nn_module)
  File "/data/users/ezyang/pytorch-tmp/torch/jit/_recursive.py", line 264, in get_or_create_concrete_type
    concrete_type_builder = infer_concrete_type_builder(nn_module)
  File "/data/users/ezyang/pytorch-tmp/torch/jit/_recursive.py", line 128, in infer_concrete_type_builder
    assert attr_type.is_interface_type()
AssertionError

----------------------------------------------------------------------
Ran 2 tests in 0.152s

FAILED (failures=1)
(/home/ezyang/local/pytorch-tmp-env) [ezyang@devvm066.ash0 ~/local/pytorch-tmp] python test/test_jit.py TestScript.test_attribute_in_init
Couldn't download test skip set, leaving all tests enabled...
.
----------------------------------------------------------------------
Ran 1 test in 0.047s

OK
```

Individually the tests pass but together they fail.

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#38211 Move all torch.nn.modules type annotations inline**

Just because the annotations are inline doesn't mean the files type
check; most of the newly annotated files have type errors and I
added exclusions for them in mypy.ini.  The payoff of moving
all of these modules inline is I can delete the relevant code
generation logic for the pyi files (which was added ignore
annotations that weren't actually relevant anymore.)
Because we aren't actually typechecking these modules in most
cases, it is inevitable that some of these type annotations are wrong.
I slavishly copied the old annotations from the pyi files unless there
was an obvious correction I could make.  These annotations will probably
need fixing up later.

Moving these annotations inline was really hairy because of interactions
with JIT, and also the fact that Python type erasure is a lie (inheriting
from Generic *does* change the behavior of your object). Here is
the list of things I had to fix and/or work around:

- The quantization translation passes previously barfed if the weight/bias arguments were inferred to be Optional. Previously, TorchScript type inference would have inferred that these arguments were non-Optional (because type inference happens after module construction), but accurate type annotations on these parameters override this inference process, causing the arguments to be optional. I fixed this by making the quantized operator signatures line up exactly with the non-quantized signatures, so we never change the types of the arguments. This change involved mostly making a bunch of quantized kernels take optional, and then error if they were passed nullopt. (You can have any color you like, as long as it's non-null.)
- I removed Generic support for Module and ModuleList. The intentions behind this were admirable, but making Module inherit from Generic ended up being more headache than it was worth. First, in Python 3.6 and earlier, Generic has a nontrivial metaclass, which means all subsequent metaclass shenanigans (e.g., ScriptModule) need to line up the right metaclass. Second, Generic defines `__new__` specially, which means that `inspect.signature` doesn't work (see https://bugs.python.org/issue40897), and I found a case of people using precisely this in the wild. Between these two problems, and also the general problem which is that the parametrization here is an incomplete fix (parametrization helps with output typing, but it doesn't solve problems with input typing (and with mypy as it stands this is unfixable, see https://github.com/python/mypy/issues/3028) I decided to just eliminate Module generics entirely. We still apply the Callable trick so that subclasses of Module don't cause mypy to complain, but otherwise you are on your own for getting accurate type information out of Modules.
- The `Callable` trick on `forward` caused TorchScript to stop performing inference on the forward body, which is bad because in general we can only figure out the most accurate type by doing TorchScript inference. I added a special case to `infer_type` to ensure we always do inference for `Module.forward`, even if it is annotated (which it is), and another special case to make sure we ignore references to Callable (which we shouldn't process) recursively.
- When `__annotations__` is set on a class (as is the case when you add type annotations), JIT will incorrectly add further annotations to the parent class. This PR fixes #39463 by testing if `__annotations__` is defined on the specific class, excluding parent classes from the test.
- Added a missing fake source range to the invocation of `get_signature`
- In some cases, we cannot provide accurate typing for parameters on modules. This usually occurs when you have an `Optional[Tensor]` parameter, whose optional-ness is determined at `__init__` time. Without the annotation, TorchScript will infer the correct refined type depending on arguments to the constructor, but with the annotation, it will never do a refinement at `__init__` time, and you'll end up with the wrong type. I ended up just straight up deleting type annotations in all of these cases. A more robust fix might be to make some way to force TorchScript to do inference even if there is an explicit annotation, in case of refinement.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21497397](https://our.internmc.facebook.com/intern/diff/D21497397)

### CODE REVIEW

**Comment by rgommers** [comment]
There's a lot of failures that end in
```
...
May 11 02:45:31   File "/opt/conda/lib/python3.6/site-packages/torch/jit/_recursive.py", line 86, in infer_type
May 11 02:45:31     attr_type = torch.jit.annotations.ann_to_type(class_annotations[name], _jit_internal.fake_range())
May 11 02:45:31   File "/opt/conda/lib/python3.6/site-packages/torch/jit/annotations.py", line 298, in ann_to_type
May 11 02:45:31     raise ValueError("Unknown type annotation: '{}'".format(ann))
May 11 02:45:31 ValueError: Unknown type annotation: '<class 'torch.nn.parameter.Parameter'>'
```
See for example `pytorch_linux_xenial_py3_6_gcc5_4_test`.

**Comment by rgommers** [comment]
Had a look at `parameter.pyi`, that looks perfectly fine. Probably a known issue of or missing feature of the JIT?

File: torch/jit/__init__.py [hunk 1]
```diff
@@ -28,11 +28,22 @@
...
+    from typing import GenericMeta
+except ImportError:
+    GenericMeta = type
+
 
 # These are imported so users can access them from the `torch.jit` module
 from torch._jit_internal import Final, _overload, _overload_method
 from torch._jit_internal import ignore, export, unused
 
+T_co = TypeVar('T_co', covariant=True)
```
Comment (rgommers) [inline torch/jit/__init__.py:45]: This isn't used in this file, so could use a comment about why it needs to be here.

Comment (ezyang) [reply torch/jit/__init__.py:45]: Dead code, should be deleted

**Comment by ezyang** [comment]
RIP.

```
======================================================================
FAIL: test_circular_dependency (__main__.TestScript)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/test_jit.py", line 5440, in test_circular_dependency
    self.getExportImportCopy(C())
  File "/data/users/ezyang/pytorch-tmp/torch/jit/__init__.py", line 1528, in init_then_script
    original_init(self, *args, **kwargs)
  File "test/test_jit.py", line 5433, in __init__
    self.foo = torch.nn.Sequential(B())
  File "/data/users/ezyang/pytorch-tmp/torch/jit/__init__.py", line 1534, in init_then_script
    self.__dict__["_actual_script_module"] = torch.jit._recursive.create_script_module(self, make_stubs)
  File "/data/users/ezyang/pytorch-tmp/torch/jit/_recursive.py", line 297, in create_script_module
    concrete_type = concrete_type_store.get_or_create_concrete_type(nn_module)
  File "/data/users/ezyang/pytorch-tmp/torch/jit/_recursive.py", line 256, in get_or_create_concrete_type
    concrete_type_builder = infer_concrete_type_builder(nn_module)
  File "/data/users/ezyang/pytorch-tmp/torch/jit/_recursive.py", line 120, in infer_concrete_type_builder
    assert attr_type.is_interface_type()
AssertionError
```

I don't know how I'm going to solve this. Also, it doesn't repro if I run only the test; I have to run the entire test suite before it.

**Comment by rgommers** [comment]
Hmm that looks bad. Maybe it's fixed by not moving the stubs for `container.py` inline? The stubs for `module.py` and `container.py` are 2 out of ~25 stub files, that would still make sense.

**Comment by ezyang** [comment]
I successfully found the pair of tests that is sufficient to trigger this error #39463 and am debugging more

**Comment by ezyang** [comment]
I killed a bunch more of the errors but this is going to require JIT review now.

**Comment by ezyang** [comment]
@jerryzh168 I added you for the quantization changes. These changes make arguments to quantized batch norm and normalization optional, lining them up with ATen, which is necessary to avoid internal assert failures in JIT. The situation is kind of subtle, but essentially if I give an accurate annotation for the parameters on these classes (optional), then the quantization pass chokes because it is looking for a Tensor? schema for the quantized version, but there is only a Tensor schema. These changes should be semantics preserving. 

@suo I added you for all the JIT changes

These changes probably can be factored into separate PRs but I haven't done so yet as I'm still squashing bugs (and they might be more difficult to test on their own)

File: aten/src/ATen/native/quantized/cpu/qnormalization.cpp [hunk 2]
```diff
@@ -123,33 +123,43 @@ TORCH_LIBRARY_IMPL(quantized, QuantizedCPU, m) {
   m.impl("layer_norm", [](
     Tensor input,
     std::vector<int64_t> normalized_shape,  // because IntArrayRef doesn't work
-    Tensor weight /* optional */,
-    Tensor bias /* optional */,
+    c10::optional<Tensor> weight,
+    c10::optional<Tensor> bias /* optional */,
```
Comment (jerryzh168) [inline aten/src/ATen/native/quantized/cpu/qnormalization.cpp:127]: nit: this comment can be removed as well

**Comment by raghuramank100** [comment]
Do you also plan to change the type annotations for modules in nn/quantized, nn/intrinsic and nn/qat?

**Comment by ezyang** [comment]
> Do you also plan to change the type annotations for modules in nn/quantized, nn/intrinsic and nn/qat?

Not specifically :) But once this PR goes in it should be easy for anyone to go in and do those updates. This PR was hard because I had to fix a bunch of JIT bugs while I was at it.

**Comment by ezyang** [comment]
Interesting unintended consequence of this PR. fairseq does this:

```
def with_incremental_state(cls):
    cls.__bases__ = (FairseqIncrementalState,) + tuple(b for b in cls.__bases__ if b != FairseqIncrementalState)
    return cls
```

However, apparently Generic is not the same object layout as object, and so this `__bases__` assignment is illegal with the error of:

```
> TypeError: __bases__ assignment: 'FairseqIncrementalState' object layout differs from 'Module'
```

It can be made to work by making FairseqIncrementalState inherit from Generic but it leaves a bad taste in my mouth.

**Comment by ezyang** [comment]
I give up on making Module Generic. The straw that broke the camel's back was https://bugs.python.org/issue40897 If people were using this in their own typed code bases I apologize; maybe some intrepid users can maintain a typeshed stub that has fancier types for PyTorch. Doesn't seem like a well supported usecase in mypy though.

**Comment by ezyang** [comment]
Description is updated with all of the new stuff I had to do.

**Comment by suo** [comment]
JIT changes look okay to me, although a little tragic 😛

File: torch/nn/modules/conv.py [hunk 3]
```diff
@@ -19,6 +21,19 @@ class _ConvNd(Module):
...
+    transposed: bool
+    output_padding: Tuple[int, ...]
+    groups: int
+    padding_mode: str
+    weight: Tensor
+    bias: Optional[Tensor]
+
     def __init__(self, in_channels, out_channels, kernel_size, stride,
                  padding, dilation, transposed, output_padding,
                  groups, bias, padding_mode):
```
Comment (malfet) [inline torch/nn/modules/conv.py:39]:
```suggestion
    def __init__(self, 
                 in_channels: int,
                 out_channels: int,
                 kernel_size: _size_1_t,
                 stride: _size_1_t,
                 padding: _size_1_t,
                 dilation: _size_1_t,
                 transposed: bool,
                 output_padding: _size_1_t,
                 groups: int,
                 bias: Optional[Tensor],
                 padding_mode: str) -> None:
```

File: torch/nn/modules/distance.py [hunk 4]
```diff
@@ -26,14 +28,17 @@ class PairwiseDistance(Module):
         >>> output = pdist(input1, input2)
     """
     __constants__ = ['norm', 'eps', 'keepdim']
+    norm: float
+    eps: float
+    keepdim: bool
 
-    def __init__(self, p=2., eps=1e-6, keepdim=False):
+    def __init__(self, p: float = 2., eps: float = 1e-6, keepdim: bool = False):
```
Comment (malfet) [inline torch/nn/modules/distance.py:35]:
```suggestion
    def __init__(self, p: float = 2., eps: float = 1e-6, keepdim: bool = False) -> None:
```

File: torch/nn/modules/distance.py [hunk 5]
```diff
@@ -58,11 +63,13 @@ class CosineSimilarity(Module):
         >>> output = cos(input1, input2)
     """
     __constants__ = ['dim', 'eps']
+    dim: int
+    eps: float
 
-    def __init__(self, dim=1, eps=1e-8):
+    def __init__(self, dim: int = 1, eps: float = 1e-8):
```
Comment (malfet) [inline torch/nn/modules/distance.py:69]:
```suggestion
    def __init__(self, dim: int = 1, eps: float = 1e-8) -> None:
```

File: torch/nn/modules/dropout.py [hunk 6]
```diff
@@ -1,19 +1,23 @@
...
+from torch import Tensor
+
 
 class _DropoutNd(Module):
     __constants__ = ['p', 'inplace']
+    p: float
+    inplace: bool
 
-    def __init__(self, p=0.5, inplace=False):
+    def __init__(self, p: float = 0.5, inplace: bool = False):
```
Comment (malfet) [inline torch/nn/modules/dropout.py:12]:
```suggestion
    def __init__(self, p: float = 0.5, inplace: bool = False) -> None:
```

**Review by malfet — APPROVED** [review]

**Pull request closed without merging**